drivers/volume/azure/azure.go

```diff
@@ -0,0 +1,432 @@
+		}
+	}
+
+	if provisioner == "" {
+		// Try to get info from the PV since storage class could be deleted
+		pv, err := k8s.Instance().GetPersistentVolume(pvc.Spec.VolumeName)
+		if err != nil {
+			logrus.Warnf("Error getting pv %v for pvc %v: %v", pvc.Spec.VolumeName, pvc.Name, err)
+			return false
+		}
+		return a.OwnsPV(pv)
+	}
+
+	if provisioner != provisionerName &&
+		!isCsiProvisioner(provisioner) {
+		logrus.Debugf("Provisioner in Storageclass not Azure: %v", provisioner)
+		return false
+	}
+	return true
+}
+
+func (a *azure) OwnsPV(pv *v1.PersistentVolume) bool {
+	var provisioner string
+	// Check the annotation in the PV for the provisioner
+	if val, ok := pv.Annotations[pvProvisionedByAnnotation]; ok {
+		provisioner = val
+	} else {
+		// Finally check the volume reference in the spec
+		if pv.Spec.AzureDisk != nil {
+			return true
+		}
+	}
+	if provisioner != provisionerName &&
+		!isCsiProvisioner(provisioner) {
+		logrus.Debugf("Provisioner in Storageclass not AzureDisk: %v", provisioner)
+		return false
+	}
+	return true
+}
+
+func isCsiProvisioner(provisioner string) bool {
+	return false
+}
+
+func (a *azure) StartBackup(backup *storkapi.ApplicationBackup,
+	pvcs []v1.PersistentVolumeClaim,
+) ([]*storkapi.ApplicationBackupVolumeInfo, error) {
+	volumeInfos := make([]*storkapi.ApplicationBackupVolumeInfo, 0)
+
+	for _, pvc := range pvcs {
+		if pvc.DeletionTimestamp != nil {
+			log.ApplicationBackupLog(backup).Warnf("Ignoring PVC %v which is being deleted", pvc.Name)
+			continue
+		}
+		volumeInfo := &storkapi.ApplicationBackupVolumeInfo{}
+		volumeInfo.PersistentVolumeClaim = pvc.Name
+		volumeInfo.Namespace = pvc.Namespace
+		volumeInfo.DriverName = driverName
+		volumeInfos = append(volumeInfos, volumeInfo)
+
+		pvName, err := k8s.Instance().GetVolumeForPersistentVolumeClaim(&pvc)
+		if err != nil {
+			return nil, fmt.Errorf("error getting PV name for PVC (%v/%v): %v", pvc.Namespace, pvc.Name, err)
+		}
+		pv, err := k8s.Instance().GetPersistentVolume(pvName)
+		if err != nil {
+			return nil, fmt.Errorf("error getting pv %v: %v", pvName, err)
+		}
+		volume := pv.Spec.AzureDisk.DiskName
+		disk, err := a.diskClient.Get(context.TODO(), a.resourceGroup, volume)
+		if err != nil {
+			return nil, err
+		}
+		volumeInfo.Volume = pvc.Spec.VolumeName
+		snapshot := compute.Snapshot{
+			Name: to.StringPtr("stork-snapshot-" + string(uuid.NewUUID())),
+			SnapshotProperties: &compute.SnapshotProperties{
+				CreationData: &compute.CreationData{
+					CreateOption:     compute.Copy,
+					SourceResourceID: disk.ID,
+				},
+			},
+			Tags: map[string]*string{
+				"created-by":           to.StringPtr("stork"),
+				"backup-uid":           to.StringPtr(string(backup.UID)),
+				"source-pvc-name":      to.StringPtr(pvc.Name),
+				"source-pvc-namespace": to.StringPtr(pvc.Namespace),
+			},
+			Location: disk.Location,
+		}
+		_, err = a.snapshotClient.CreateOrUpdate(context.TODO(), a.resourceGroup, *snapshot.Name, snapshot)
```

Comment:
do we need to call cloud api using backoff or should we wait for reconsiler in case of failure

Reply:
All the calls are async, so this will just trigger the snapshot and return. We'll check again during resync if the snapshot is complete. If this call fails we'll fail the ApplicationBackup.